### PR TITLE
Support Ktor Engine

### PR DESCRIPTION
**Pull request by yt8492**
Add to apollo-runtime/commonMain: 
- KtorHttpEngine
- KtorWebsocketEngine
- Extension for builder

implements for #3796

### CODE REVIEW

**Comment by BoD** [comment]
Thanks a lot for the contribution! 🙏

A few thoughts:

- We will want to move this to a dedicated module, to avoid having `apollo-runtime` depending on Ktor
- It would be nice if we could reuse `KtorHttpEngine` /  `KtorWebSocketEngine`for Js instead of it being duplicated (not 100% sure how to do this - it depends on how we proceed for the 1st item above)
- We may need a few more tests (in particular around subscriptions/websockets)

We can help or take care of these of course!

**Comment by martinbonnin** [comment]
> It would be nice if we could reuse KtorHttpEngine / KtorWebSocketEnginefor Js instead of it being duplicated

I've been thinking about this, it's a pain because of a circular module dependency :(. `HttpEngine` is defined in `apollo-runtime`. So if we want to have `apollo-runtime` depend on `apollo-engine-ktor`, we also need `apollo-engine-ktor` to depend on `apollo-runtime` :/. We _could_ introduce a `apollo-runtime-api` but it's quite the change. I'd prefer duplicating the implementations for the time being.

File: gradle/libraries.toml [hunk 1]
```diff
@@ -30,11 +30,11 @@ kotlin-plugin-min = "1.6.0"
 kotlin-plugin = "1.9.0"
 kotlin-plugin-max = "1.9.0"
 kotlin-stdlib = "1.6.21"
-kotlinx-coroutines = "1.6.4"
+kotlinx-coroutines = "1.7.3"
 kotlinx-datetime = "0.4.0"
 kotlinx-serialization-runtime = "1.5.0"
 ksp = "1.9.0-1.0.11"
-ktor = "2.2.2"
+ktor = "2.3.2"
```
Comment (BoD) [inline gradle/libraries.toml:37]: Turns out 2.3.3 was just released
```suggestion
ktor = "2.3.3"
```

File: libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/KtorHttpEngine.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import io.ktor.http.HttpHeaders
+import io.ktor.util.flattenEntries
+import okio.Buffer
+import kotlin.coroutines.cancellation.CancellationException
+
+class KtorHttpEngine(
+    private val client: HttpClient,
+): HttpEngine {
+
+  var disposed = false
```
Comment (BoD) [inline libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/http/KtorHttpEngine.kt:23]: Can probably be private

**Comment by yt8492** [comment]
@BoD 
Thank you for reviewing my changes! I applied your suggestion to this pull request.

**Comment by BoD** [comment]
Thank you so much @yt8492, this looks good!

Meanwhile, I've found a few issues with the WebSocketEngine that made a few tests fails, but it looks like I just managed to fix the last one. I will rebase my changes on your branch and push them here.

**Review by BoD — APPROVED** [review]
👍

**Review by martinbonnin — APPROVED** [review]
Thanks!

**Pull request merged**